### PR TITLE
Lint: Lint for std::for_each_n 

### PR DESCRIPTION
**Pull request by KBaichoo**
Signed-off-by: Kevin Baichoo <kbaichoo@google.com>

For an explanation of how to fill out the fields, please see the relevant section
in [PULL_REQUESTS.md](https://github.com/envoyproxy/envoy/blob/main/PULL_REQUESTS.md)

Commit Message: Lint: Lint for std::for_each_n 
Additional Description: It's insufficiently supported in C++17. See #14923
Risk Level: low
Testing: unit test
Docs Changes: NA
Release Notes: NA
Platform Specific Features: NA

### CODE REVIEW

File: source/common/http/http2/codec_impl.cc [hunk 1]
```diff
@@ -1409,8 +1409,12 @@ void ConnectionImpl::dumpState(std::ostream& os, int indent_level) const {
...
 
   os << spaces << "Number of active streams: " << active_streams_.size() << " Active Streams:\n";
-  std::for_each_n(active_streams_.begin(), std::min<size_t>(active_streams_.size(), 100),
-                  [&](auto& stream) { DUMP_DETAILS(stream); });
+  const size_t num_iterations = std::min<size_t>(active_streams_.size(), 100);
+  auto stream = active_streams_.begin();
+  for (size_t i = 0; i < num_iterations; ++i) {
+    DUMP_DETAILS(*stream);
+    ++stream;
+  }
```
Comment (antoniovicente) [inline source/common/http/http2/codec_impl.cc:1417]: Consider:

```
size_t count = 0;
for (auto& stream : active_streams_) {
  DUMP_DETAILS(stream);
  if (++count >= 100) {
    break;
  }
}
```

**Review by antoniovicente — COMMENTED** [review]
Thanks for addressing the compile issues introduced by #14923

Change looks good if we decide to avoid std::for_each_n for the time being.

**Comment by antoniovicente** [comment]
cc @jrajahalme

File: source/common/http/http2/codec_impl.cc (hunk 1 above)
Comment (KBaichoo) [reply source/common/http/http2/codec_impl.cc:1418]: Done.

**Comment by lizan** [comment]
I'm generally not in favor of this kind of limitation because of it is not well supported in older version of stdlib (libstdc++ <= 8). We can bump build requirement if it doesn't cause a lot of churn.

**Comment by jrajahalme** [comment]
@lizan I agree that limiting use of a standardized C++17 feature is not ideal, but `for_each_n` is available only on the newest libc++. However, I'd like to retain compatibility with GCC on Ubuntu 18.04 for now. GCC due to working cross-compilation toolchain we have for arm64 (I have no idea how to do that on clang), and Ubuntu 18.04 for ability to run the same Envoy binary in Ubuntu 18.04 and 20.04 based images. Unfortunately `for_each_n` does not work on GCC 7 that is the system compiler for Ubuntu 18.04.

**Comment by jrajahalme** [comment]
To clarify, bumping build requirements is a potential concern for projects building on Envoy, not just Envoy itself. There is already some churn due to Envoy CI apparently using only a bleeding edge clang setup :-)

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**